packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/data/TerariumAssetService.java

```diff
@@ -127,6 +140,41 @@ public Optional<T> updateAsset(final T asset) throws IOException, IllegalArgumen
 
 		asset.setUpdatedOn(Timestamp.from(Instant.now()));
 		elasticService.index(getAssetIndex() , asset.getId().toString(), asset);
+
+		// Update the related ProjectAsset.assetName
+		final Optional<ProjectAsset> projectAsset = projectAssetService.getProjectAssetById(asset.getId());
+		if (projectAsset.isPresent()) {
+			final String assetName = getAssetName(asset);
+			if (assetName != null) {
+				projectAsset.get().setAssetName(assetName);
+				projectAssetService.save(projectAsset.get());
+			} else {
+				log.info("Could not update the project asset name for asset with id: " + asset.getId() + " because the asset name is null");
+			}
+		}
+
 		return Optional.of(asset);
 	}
+
+	/**
+	 * Get the name of an asset
+	 * @param asset
+	 * @return assetName
+	 * @param <T>
+	 */
+	private static <T> String getAssetName(T asset) {
+		// Force cast to the correct type because Terrarium Asset does not provide a bloody name
+		if (asset instanceof Dataset) {
+			return ((Dataset) asset).getName();
+		} else if (asset instanceof Model) {
+			return ((Model) asset).getHeader().getName();
+		} else if (asset instanceof Workflow) {
+			return ((Workflow) asset).getName();
+		} else if (asset instanceof Code) {
+			return ((Code) asset).getName();
+		} else if (asset instanceof DocumentAsset) {
+			return ((DocumentAsset) asset).getName();
+		}
+		return null;
+	}
```

Comment:
I don't think you are really using the generics `<T>` if that is the intent here, as far as I remember you can do instanceof without needing the <T> part.

Reply:
Removed and changed to `Object`